File: 4-7.cpp
#include "iostream"
#include "opencv2/opencv.hpp"   

using namespace std;
using namespace cv;

static void on_trackbar(int, void*){

}
int main(){
    Mat img_color = imread("red-ball.jpeg", IMREAD_COLOR);
    
    if(img_color.empty()){
        cout << "can not read image file" << endl;
        return -1;
    }
    imshow("color", img_color);
    Mat img_gray;
    cvtColor(img_color, img_gray, COLOR_BGR2GRAY);
    imshow("GraySclae", img_gray);

    namedWindow("Binary");
    createTrackbar("threshold", "Binary", 0, 255, on_trackbar);
    setTrackbarPos("threshold", "Binary", 127);

    while(1){
        int thresh = getTrackbarPos("threshold", "Binary");

        Mat img_binary;
        threshold(img_gray, img_binary, thresh, 255, THRESH_BINARY_INV);

        imshow("Binary", img_binary);

        if(waitKey(1) == 27)
            break;
    }
    return 0;
}

File: 4-7.py
import imghdr
import cv2 as cv
import sys

from cv2 import IMREAD_COLOR

def on_trackbar(x):
    pass

img_color = cv.imread('red-ball.jpeg', IMREAD_COLOR)

if img_color is None:
    print('can not read image file')
    sys.exit(1)
cv.imshow('Color', img_color)

img_gray = cv.cvtColor(img_color, cv.COLOR_BGR2GRAY)
cv.imshow('GrayScale', img_gray)

cv.namedWindow('Binary')    
cv.createTrackbar('threshold', 'Binary', 0,255, on_trackbar)
cv.setTrackbarPos('threshold', 'Binary', 127)

while True:
    thresh = cv.getTrackbarPos('threshold', 'Binary')

    retval, img_binary = cv.threshold(img_gray, thresh, 255, cv.THRESH_BINARY_INV)

    cv.imshow('Binary', img_binary)
    if cv.waitKey(1) & 0xFF == 27:
        break